### PR TITLE
Roles

### PR DESCRIPTION
**Pull request by fergusdixon**
Added roles middleware

### CODE REVIEW

File: app/Http/Controllers/RolesController.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+namespace App\Http\Controllers;
+
+use App\Role;
+use App\User;
+use Illuminate\Support\Facades\Auth;
+use Webpatser\Uuid\Uuid;
+
+class RolesController extends Controller
+{
```
Comment (darrynten) [inline app/Http/Controllers/RolesController.php:11]: we also need `getRoles` to list all the roles in the system

File: app/Http/Controllers/RolesController.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    public function getUserRoles($id)
+    {
```
Comment (darrynten) [inline app/Http/Controllers/RolesController.php]: This should be in the users controller

File: app/Http/Controllers/RolesController.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    public function getRole($name)
+    {
+        return Role::where('name', $name)->first();
+    }
+
+    public function assignRole($id, $role)
+    {
+        User::where('id', $id)->first()->assignRole($role);
+        return response('OK', 200);
```
Comment (darrynten) [inline app/Http/Controllers/RolesController.php]: User controller

File: app/Http/Controllers/RolesController.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    public function revokeRole($id, $role)
+    {
+        User::where('id', $id)->first()->revokeRole($role);
+        return response('OK', 200);
```
Comment (darrynten) [inline app/Http/Controllers/RolesController.php]: Again

File: app/Http/Middleware/RolesMiddleware.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+     *
+     * @param  \Illuminate\Http\Request $request
+     * @param  \Closure $next
+     * @param array $requiredRoles
+     * @return mixed
+     */
+    public function handle($request, Closure $next, ...$requiredRoles)
+    {
+        if (empty($requiredRoles)) {
+            if (Auth::user()->roles->count() != 0) {
```
Comment (darrynten) [inline app/Http/Middleware/RolesMiddleware.php]: `!==` _never_ `!=` (except for, like, 0.00001% of the time)

File: app/Http/Middleware/RolesMiddleware.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+                return $next($request);
+            }
+        }
+
+        foreach ($requiredRoles as $role) {
+            if (empty(Role::where('name', $role)->first())) {
+                return response()->json(['error' => 'Undefined role on route'], 500);
+            }
```
Comment (darrynten) [inline app/Http/Middleware/RolesMiddleware.php]: space between these 2 ifs

File: app/Role.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\SoftDeletes;
+
+class Role extends Model
+{
+    use SoftDeletes;
+
+    protected $attributes = [
+        'active' => true //Roles are active on creation
```
Comment (darrynten) [inline app/Role.php]: No inline comments

File: database/migrations/2017_12_08_142555_add_roles_tables.php [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('roles', function (Blueprint $table) {
+            $table->bigIncrements('id');
+            $table->uuid('_id')->unique();
+            $table->string('name')->unique();
+            $table->boolean('active');
```
Comment (darrynten) [inline database/migrations/2017_12_08_142555_add_roles_tables.php]: is_active

Comment (darrynten) [reply database/migrations/2017_12_08_142555_add_roles_tables.php]: `$table->boolean('active')->default(true)`

File: database/seeds/DatabaseSeeder.php [hunk 9]
```diff
@@ -12,5 +12,6 @@ class DatabaseSeeder extends Seeder
     public function run()
     {
         $this->call('Users');
+        $this->call(Roles::class);
```
Comment (darrynten) [inline database/seeds/DatabaseSeeder.php]: Just `$this->call('Roles')`

File: database/seeds/Roles.php [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+        if (app()->environment('local')) {
+            DB::table('user_role')->delete();
+        }
+
+        UserRole::create([
+            'user_id' => 1,
+            'role_id' => 1,
+        ]);
+        UserRole::create([
```
Comment (darrynten) [inline database/seeds/Roles.php]: Space between these 2 creates

File: routes/web.php [hunk 11]
```diff
@@ -67,6 +69,18 @@
...
+    $router->group(['prefix' => 'roles', 'middleware' => ['roles:admin']], function () use ($router) {
+        $router->get('/{id}', 'RolesController@getUserRoles');
+        $router->get('/getRole/{name}', 'RolesController@getRole');
+        $router->post('/assignRole/{id}/{role}', 'RolesController@assignRole');
+        $router->post('/revokeRole/{id}/{role}', 'RolesController@revokeRole');
+        $router->post('/createRole/{name}', 'RolesController@createRole');
+        $router->post('/deleteRole/{name}', 'RolesController@deleteRole');
+        $router->post('/activate/{name}', 'RolesController@activateRole');
+        $router->post('/deactivate/{name}', 'RolesController@deactivateRole');
+
```
Comment (darrynten) [inline routes/web.php]: Extraneous newline

File: tests/app/Http/AuthControllerTest.php [hunk 12]
```diff
@@ -181,7 +181,7 @@ public function testChangePasswordSamePassword()
             'password' => 'password',
             'newpassword' => 'password',
         ], [
-            'Debug-Token' => env('DEBUG_KEY'),
+            //'Debug-Token' => env('DEBUG_KEY'),
```
Comment (darrynten) [inline tests/app/Http/AuthControllerTest.php]: No commented out code

File: tests/app/Http/Middleware/RoleMiddlewareTest.php [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+        parent::setUp();
+        $this->post('/login', [
+            'username' => 'admin',
+            'password' => 'admin',
+        ], [ 'Debug-Token' => env('DEBUG_TOKEN')]);
+    }
+
+    public function testBadRole()
+    {
+        //before adding admin role
```
Comment (darrynten) [inline tests/app/Http/Middleware/RoleMiddlewareTest.php]: space between `//` and comment

File: tests/app/Http/Middleware/RoleMiddlewareTest.php [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+    public function testCreateRole()
+    {
+        $this->actingAs(Auth::user())->post('/roles/createRole/intern');
+
+        $this->actingAs(Auth::user())->post('/roles/assignRole/2/intern'); //Assign new role
+        $this->actingAs(Auth::user())->get('/roles/2');
+        $roles = json_decode($this->response->getContent());
+        $this->assertEquals('intern', $roles[0]->name);
+
```
Comment (darrynten) [inline tests/app/Http/Middleware/RoleMiddlewareTest.php]: No newline

File: tests/app/Http/Middleware/RoleMiddlewareTest.php [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+    {
+        $this->assertEquals(1, Role::where('name', 'admin')->first()->active);
+
+        $this->actingAs(Auth::user())->get('/roles/3');
+        $this->assertResponseStatus(200);
+
+        $this->actingAs(Auth::user())->post('/roles/deactivate/admin');
+
+//        $this->actingAs(Auth::user())->get('/roles/3');
+//        $this->assertResponseStatus(401);
```
Comment (darrynten) [inline tests/app/Http/Middleware/RoleMiddlewareTest.php]: Remove

File: tests/app/Http/Middleware/RoleMiddlewareTest.php [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+        });
+        $this->actingAs(Auth::user())->post('/roles/createRole/intern'); //Admin user
+        $this->actingAs(Auth::user())->post('/roles/createRole/system');//Admin user
+
+        //Fail
+        $this->actingAs(User::where('id', 2)->first())->get('/test');
+        $this->assertResponseStatus(401);
+        $this->assertEquals('{"error":"Incorrect Role"}', $this->response->getContent());
+
+        //Only 1st role
```
Comment (darrynten) [inline tests/app/Http/Middleware/RoleMiddlewareTest.php]: Spaces on these comment markers

**Review by darrynten — CHANGES_REQUESTED** [review]
Looking great so far :D

File: app/Http/Controllers/RolesController.php (hunk 1 above)
Comment (darrynten) [reply app/Http/Controllers/RolesController.php:11]: and getUsersForRole

File: .gitignore [hunk 17]
```diff
@@ -5,3 +5,4 @@ Homestead.yaml
 .env
 clover.xml
 tests/coverage
+composer.lock
```
Comment (darrynten) [inline .gitignore]: This file should not be ignored

File: app/BaseModel.php [hunk 18]
```diff
@@ -0,0 +1,56 @@
+<?php
+
+namespace App;
+
+use Illuminate\Database\Eloquent\Model;
+
+class BaseModel extends Model
+{
+    use ValidationTrait;
```
Comment (darrynten) [inline app/BaseModel.php:9]: Space between the use and the next function

File: app/BaseModel.php [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+    public static function loadFromUuid($uuid)
+    {
+        $class  = get_called_class();
```
Comment (darrynten) [inline app/BaseModel.php]: AFAIK his isn't the correct way of making a copy of the class that's called.

Try `__CLASS__`

File: app/Http/Controllers/RolesController.php [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    public function activateRole($name)
+    {
+        $role = $this->getRole($name);
+        $role->is_active = true;
+        $role->save();
+    }
+
+    public function getUserForRole($name)
```
Comment (darrynten) [inline app/Http/Controllers/RolesController.php]: Users (plural)

File: app/Http/Middleware/RolesMiddleware.php [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+     *
+     * @param  \Illuminate\Http\Request $request
+     * @param  \Closure $next
+     * @param array $requiredRoles
+     * @return mixed
+     */
+    public function handle($request, Closure $next, ...$requiredRoles)
+    {
+        if (empty($requiredRoles)) {
+            if (Auth::user()->roles->count() !== 0) {
```
Comment (darrynten) [inline app/Http/Middleware/RolesMiddleware.php]: Shouldn't this be Auth::user()->roles->exists() (or similar)?

Comment (darrynten) [reply app/Http/Middleware/RolesMiddleware.php]: Perhaps hasAnyRole([]) with an array of roles the method loops through and if the array is empty it checks all roles

File: app/Http/Middleware/RolesMiddleware.php [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+                return $next($request);
+            }
+        }
+
+        foreach ($requiredRoles as $role) {
+            $model = Role::loadRoleFromName($role);
```
Comment (darrynten) [inline app/Http/Middleware/RolesMiddleware.php]: Does the called method do all the validation?

File: database/migrations/2017_12_08_142555_add_roles_tables.php [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('roles', function (Blueprint $table) {
+            $table->bigIncrements('id');
+            $table->uuid('_id')->unique();
+            $table->string('name')->unique();
+            $table->boolean('is_active');
```
Comment (darrynten) [inline database/migrations/2017_12_08_142555_add_roles_tables.php]: ->default(true)

File: tests/app/Http/Middleware/RoleMiddlewareTest.php [hunk 24]
```diff
@@ -0,0 +1,221 @@
...
+        $router = $this->app->router;
+        $this->app->router->group(['middleware' => ['roles:system,intern']], function () use ($router) {
+            $router->get('/test', function () {
+                return 'Test';
+            });
+        });
+        $this->actingAs(Auth::user())->post('/roles/createRole/intern');
+        $this->actingAs(Auth::user())->post('/roles/createRole/system');
+
+        //Fail
```
Comment (darrynten) [inline tests/app/Http/Middleware/RoleMiddlewareTest.php]: Comments need space after //

File: routes/web.php [hunk 25]
```diff
@@ -67,6 +69,19 @@
         $router->post('/reset/{token}', 'ResetController@postReset');
     });
 
+    $router->group(['prefix' => 'roles', 'middleware' => ['roles:admin']], function () use ($router) {
+        $router->get('/getUserRoles/{userId}', 'UserController@getUserRoles');
+        $router->get('/getUsers/{role}', 'RolesController@getUserForRole');
+        $router->get('/getRole/{name}', 'RolesController@getRole');
+        $router->get('/getAllRoles', 'RolesController@getRoles');
+        $router->post('/assignRole/{userId}/{role}', 'UserController@assignRole');
```
Comment (darrynten) [inline routes/web.php]: Params should be role then user

File: database/seeds/Roles.php [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    public function run()
+    {
+        Model::unguard();
+
+        if (app()->environment('local')) {
+            DB::table('roles')->delete();
+        }
+
+        Role::create([
```
Comment (darrynten) [inline database/seeds/Roles.php:25]: You also need `system` and `user` roles

File: app/BaseModel.php [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+        $model = new $class;
+        $model->checkValid($uuid, $class);
+        $id = self::getIdFromUuid($uuid, $model);
+        return $class::where('id', $id)->first();
+    }
+
+    public static function loadRoleFromName($name)
```
Comment (darrynten) [inline app/BaseModel.php]: Please add  PHP docblocks to all methods

**Review by darrynten — CHANGES_REQUESTED** [review]
Looks good so far.

Some changes needed, as well as an update to the readme explaining things.

Lekker

File: app/BaseModel.php (hunk 19 above)
Comment (fergusdixon) [reply app/BaseModel.php]: http://willi.am/blog/2011/09/05/get-called-class-name-in-php/

Comment (fergusdixon) [reply app/BaseModel.php]: `__CLASS__` returns `BaseModel`

File: app/Http/Middleware/RolesMiddleware.php (hunk 22 above)
Comment (fergusdixon) [reply app/Http/Middleware/RolesMiddleware.php]: Yip

File: README.md [hunk 28]
```diff
@@ -196,6 +196,30 @@ Make sure you set the appropriate variables in your .env
...
+```
+
+##Role endpoints
+- `GET` `/roles/getUsers/{role}`
+- `GET` `/roles/getRole/{name}`
+- `GET` `/roles/getAllRoles`
+- `POST` `/roles/createRole/{name}`
+- `POST` `/roles/deleteRole/{name}`
+- `POST` `/roles/activate/{name}`
+- `POST` `/roles/deactivate/{name}`
```
Comment (darrynten) [inline README.md]: I left some notes in slack about the names of these routes

GET /roles/{name}/users
GET /roles/{name}
GET /roles

POST /roles/{name}
DELETE /roles/{name}
POST /roles/{name}/activate
POST /roles/{name}/deactivate

File: README.md [hunk 29]
```diff
@@ -196,6 +196,30 @@ Make sure you set the appropriate variables in your .env
...
+
+##Assigning roles to users
+- `GET` `/roles/getUserRoles/{userId}`
+- `POST` `/roles/assignRole/{role}/{userId}`
+- `POST` `/roles/revokeRole/{role}/{userId}`
```
Comment (darrynten) [inline README.md]: GET /users/{userId}/roles
POST /users/{userId}/roles/assign/{name}
POST /users/{userId}/roles/revoke/{name}

File: README.md (hunk 28 above)
Comment (darrynten) [reply README.md]: I'm also wondering if we should use the UUIDs here.

It's easy to change the url to assign to admin if you see an assign call uses a name. If it has a UUID there's no names to fiddle with.

Let's make these UUIDs instead of name based.

Devs can get the UUIDs from the GET /roles call

Name based is just too risky for something this sensitive

File: app/Http/Middleware/RolesMiddleware.php [hunk 30]
```diff
@@ -0,0 +1,39 @@
...
+
+use App\Http\Controllers\UserController;
+use App\Role;
+use App\User;
+use Closure;
+use Illuminate\Support\Facades\Auth;
+
+class RolesMiddleware
+{
+
```
Comment (darrynten) [inline app/Http/Middleware/RolesMiddleware.php]: Unneeded newline

File: database/seeds/Roles.php [hunk 31]
```diff
@@ -0,0 +1,66 @@
...
+            'id' => 1,
+            '_id' => Uuid::generate(4),
+            'name' => 'admin',
```
Comment (darrynten) [inline database/seeds/Roles.php]: System should be id 1.

It's very common to compromise and escalate user user id 1 as it's usually admin.

Making a really limited user that cannot login or do much as id 1 or with assigning role id 1 would make an attack that successfully escalates to user id 1 will leave the attacker with a gimped account or permission that can't do much.

File: database/seeds/Roles.php [hunk 32]
```diff
@@ -0,0 +1,66 @@
...
+        ]);
+
+        Role::create([
+            'id' => 3,
+            '_id' => Uuid::generate(4),
+            'name' => 'user',
+            'is_active' => true,
+            'created_by' => 5,
+            'updated_by' => 4,
+        ]);
```
Comment (darrynten) [inline database/seeds/Roles.php]: You can add a 4th role, developer

File: database/seeds/Roles.php [hunk 33]
```diff
@@ -0,0 +1,66 @@
...
+
+        Role::create([
+            'id' => 3,
+            '_id' => Uuid::generate(4),
+            'name' => 'user',
+            'is_active' => true,
+            'created_by' => 5,
```
Comment (darrynten) [inline database/seeds/Roles.php]: These should be marked assigned and created by the system user

File: routes/web.php [hunk 34]
```diff
@@ -67,6 +69,21 @@
...
 
+    $router->group(['prefix' => 'roles', 'middleware' => ['roles:admin']], function () use ($router) {
+
+        $router->get('/getUsers/{role}', 'RolesController@getUsersForRole');
+        $router->get('/getRole/{name}', 'RolesController@getRole');
+        $router->get('/getAllRoles', 'RolesController@getRoles');
+        $router->post('/createRole/{name}', 'RolesController@createRole');
+        $router->post('/deleteRole/{name}', 'RolesController@deleteRole');
+        $router->post('/activate/{name}', 'RolesController@activateRole');
+        $router->post('/deactivate/{name}', 'RolesController@deactivateRole');
```
Comment (darrynten) [inline routes/web.php]: See notes above about the naming of these routes

File: tests/app/Http/Middleware/RoleMiddlewareTest.php [hunk 35]
```diff
@@ -0,0 +1,224 @@
...
+
+    public function testAssignRole()
+    {
+        $this->actingAs(Auth::user())->post('/roles/assignRole/admin/' . $this->testUserId);
+
+        $this->actingAs(Auth::user())->get('/roles/getUserRoles/' . $this->testUserId);
+
+        // Check to see if the role was assigned successfully
+        $roles = json_decode($this->response->getContent());
+        $this->assertEquals('admin', $roles[0]->name);
```
Comment (darrynten) [inline tests/app/Http/Middleware/RoleMiddlewareTest.php]: Test assigning a role that doesn't exist

File: tests/app/Http/Middleware/RoleMiddlewareTest.php [hunk 36]
```diff
@@ -0,0 +1,224 @@
...
+        // Ensure we added the role correctly
+        $this->assertEquals('admin', $roles[0]->name);
+
+        // remove the role
+        $this->actingAs(Auth::user())->post('roles/revokeRole/admin/' . $this->testUserId);
+
+        $this->actingAs($testUser)->get('/roles/getUserRoles/' . $this->testUserId);
+
+        // ensure we get access denied error
+        $this->assertResponseStatus(401);
```
Comment (darrynten) [inline tests/app/Http/Middleware/RoleMiddlewareTest.php]: Test revoking a role that doesn't exist

File: tests/app/Http/Middleware/RoleMiddlewareTest.php [hunk 37]
```diff
@@ -0,0 +1,224 @@
...
+    {
+        $this->actingAs(Auth::user())->get('/roles/getAllRoles');
+        $roles = json_decode($this->response->getContent());
+        $this->assertEquals('admin', $roles[0]->name);
+
+        $this->actingAs(Auth::user())->post('/roles/createRole/intern');
+
+        $this->actingAs(Auth::user())->get('/roles/getAllRoles');
+        $roles = json_decode($this->response->getContent());
+        $this->assertEquals('intern', $roles[3]->name);
```
Comment (darrynten) [inline tests/app/Http/Middleware/RoleMiddlewareTest.php]: Test get roles for a user that is not an admin

File: tests/app/Http/Middleware/RoleMiddlewareTest.php [hunk 38]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    public function testGetUserForRole()
+    {
+        $this->actingAs(Auth::user())->get('/roles/getUsers/admin');
+        $users = json_decode($this->response->getContent());
+        $this->assertEquals('5FFA95F4-5EB4-46FB-94F1-F2B27254725B', $users[0]->_id);
```
Comment (darrynten) [inline tests/app/Http/Middleware/RoleMiddlewareTest.php]: Also test getting an empty role

File: tests/app/Http/Middleware/RoleMiddlewareTest.php [hunk 39]
```diff
@@ -0,0 +1,224 @@
...
+    {
+        $this->actingAs(Auth::user())->post('/roles/createRole/intern');
+
+        $this->actingAs(Auth::user())->post('/roles/assignRole/intern/' . $this->testUserId);
+        $this->actingAs(Auth::user())->get('/roles/getUserRoles/' . $this->testUserId);
+        $roles = json_decode($this->response->getContent());
+
+        $this->assertEquals('intern', $roles[0]->name);
+        $this->actingAs(Auth::user())->post('/roles/createRole/mm');
+        $this->assertEquals('Role name invalid', $this->response->getContent());
```
Comment (darrynten) [inline tests/app/Http/Middleware/RoleMiddlewareTest.php]: Test create role that already exists

File: tests/app/Http/Middleware/RoleMiddlewareTest.php [hunk 40]
```diff
@@ -0,0 +1,224 @@
...
+        $this->actingAs(User::where('id', 2)->first())->get('/test');
+        $this->assertResponseStatus(200);
+        $this->assertEquals('Test', $this->response->getContent());
+
+        // both roles
+        $this->actingAs(User::where('id', 3)->first())->post('/roles/assignRole/intern' . $this->testUserId);
+
+        $this->actingAs(User::where('id', 2)->first())->get('/test');
+        $this->assertResponseStatus(200);
+        $this->assertEquals('Test', $this->response->getContent());
```
Comment (darrynten) [inline tests/app/Http/Middleware/RoleMiddlewareTest.php]: Test user having 1 active and 1 inactive, and routes with neither 1 or each, and both set

Comment (darrynten) [reply tests/app/Http/Middleware/RoleMiddlewareTest.php]: Neither, 1 of each, and having both **

**Review by darrynten — CHANGES_REQUESTED** [review]

File: README.md (hunk 28 above)
Comment (fergusdixon) [reply README.md]: Sounds much better

File: tests/app/Http/Middleware/RoleMiddlewareTest.php (hunk 37 above)
Comment (fergusdixon) [reply tests/app/Http/Middleware/RoleMiddlewareTest.php]: It will fail because getRoles is under admin protection. Should it not be?

File: tests/app/Http/Middleware/RoleMiddlewareTest.php (hunk 38 above)
Comment (fergusdixon) [reply tests/app/Http/Middleware/RoleMiddlewareTest.php]: get `/roles/` will return all roles

File: README.md [hunk 41]
```diff
@@ -196,6 +196,30 @@ Make sure you set the appropriate variables in your .env
...
+- `GET` `/roles/`
+- `POST` `/roles/{name}`
+- `DELETE` `/roles/{roleId}`
+- `POST` `/roles/{name}/activate`
+- `POST` `/roles/{name}/deactivate`
+
+##Assigning roles to users
+- `GET` `/users/{userId}/roles`
+- `POST` `/users/{userId}/roles/assign/{role}`
+- `POST` `/users/{userId}/roles/revoke/{role}`
```
Comment (darrynten) [inline README.md]: for consistency put `id` inside the `{}` on each of these as it's always id

File: app/Http/Controllers/RolesController.php [hunk 42]
```diff
@@ -0,0 +1,112 @@
...
+     * @param $name
+     * @return \Illuminate\Http\Response|\Laravel\Lumen\Http\ResponseFactory
+     */
+    public function createRole(string $name)
+    {
+        $role = Role::where('name', $name)->first();
+        if (preg_match("/[a-z]{4,}/", $name) && is_null($role)) {
+            $role = new Role();
+            $role->name = $name;
+            $role->_id = Uuid::generate(4);
```
Comment (darrynten) [inline app/Http/Controllers/RolesController.php]: Uuid::generate(4)->string

File: app/Http/Controllers/RolesController.php [hunk 43]
```diff
@@ -0,0 +1,112 @@
...
+            $role->is_active = true;
+            $role->created_by = Auth::user()->id;
+            $role->updated_by = Auth::user()->id;
+            $role->save();
+            return response()->json(['uuid' => (string)$role->_id]);
```
Comment (darrynten) [inline app/Http/Controllers/RolesController.php]: You can remove the `(string)` with the change above

File: database/migrations/2017_12_08_142555_add_roles_tables.php [hunk 44]
```diff
@@ -0,0 +1,46 @@
...
+            $table->uuid('_id')->unique();
+            $table->string('name')->unique();
+            $table->boolean('is_active')->default(true);
+            $table->unsignedBigInteger('created_by');
+            $table->unsignedBigInteger('updated_by');
+            $table->timestamps();
+            $table->softDeletes();
+        });
+
+        Schema::create('user_role', function (Blueprint $table) {
```
Comment (darrynten) [inline database/migrations/2017_12_08_142555_add_roles_tables.php:27]: user_roles?

File: routes/web.php [hunk 45]
```diff
@@ -67,6 +69,23 @@
         $router->post('/reset/{token}', 'ResetController@postReset');
     });
 
+    $router->group(['prefix' => 'roles', 'middleware' => ['roles:admin']], function () use ($router) {
+
```
Comment (darrynten) [inline routes/web.php]: Unneeded newline

File: tests/app/Http/Middleware/RoleMiddlewareTest.php [hunk 46]
```diff
@@ -0,0 +1,304 @@
...
+                    $this->adminId = $role->_id;
+                    break;
+                case 'user':
+                    $this->userId = $role->_id;
+                    break;
+                case 'system':
+                    $this->systemId = $role->_id;
+
+            }
+
```
Comment (darrynten) [inline tests/app/Http/Middleware/RoleMiddlewareTest.php]: Unneeded newline

File: tests/app/Http/Middleware/RoleMiddlewareTest.php [hunk 47]
```diff
@@ -0,0 +1,304 @@
...
+        $this->actingAs(Auth::user())->get('/users/' . $this->testUserId . '/roles');
+
+        // Check to see if the role was assigned successfully
+        $roles = json_decode($this->response->getContent());
+        $this->assertEquals('admin', $roles[0]->name);
+
+        $this->actingAs(Auth::user())->post(
+            '/users/'
+            . $this->testUserId
```
Comment (darrynten) [inline tests/app/Http/Middleware/RoleMiddlewareTest.php]: `sprintf`, never `.` string concats

File: tests/app/Http/Middleware/RoleMiddlewareTest.php [hunk 48]
```diff
@@ -0,0 +1,304 @@
...
+    }
+
+    public function testCreateRole()
+    {
+        $this->actingAs(Auth::user())->post('/roles/intern');
+
+        $internId = json_decode($this->response->getContent())->uuid;
+
+        $this->actingAs(Auth::user())->post('/users/' . $this->testUserId . '/roles/assign/' . $internId);
+        $this->actingAs(Auth::user())->get('/users/' . $this->testUserId . '/roles');
```
Comment (darrynten) [inline tests/app/Http/Middleware/RoleMiddlewareTest.php]: sprintf everywhere

File: app/Http/Controllers/RolesController.php (hunk 43 above)
Comment (darrynten) [reply app/Http/Controllers/RolesController.php]: `_id` instead of `uuid`

**Review by darrynten — CHANGES_REQUESTED** [review]

File: database/migrations/2017_12_08_142555_add_roles_tables.php (hunk 44 above)
Comment (fergusdixon) [reply database/migrations/2017_12_08_142555_add_roles_tables.php:27]: Getting very weird issues with regards to the way Eloquent implies tables, the classes are `User` and `Role` so it looks for `user_role` from what I can tell

File: README.md [hunk 49]
```diff
@@ -196,6 +196,30 @@ Make sure you set the appropriate variables in your .env
...
+middleware => ['roles']
+```
+
+##Role endpoints
+- `GET` `/roles/{roleId}/users`
+- `GET` `/roles/{Id}`
+- `GET` `/roles/`
+- `POST` `/roles/{name}`
+- `DELETE` `/roles/{roleId}`
+- `POST` `/roles/{name}/activate`
```
Comment (darrynten) [inline README.md]: These are {id} not name

File: app/BaseModel.php [hunk 50]
```diff
@@ -0,0 +1,38 @@
...
+     *
+     * @param $uuid
+     * @return mixed
+     */
+    public static function loadFromUuid($uuid)
+    {
+        $class  = get_called_class();
+        $model = new $class;
+        $model->checkValid($uuid, $class);
+        $classId = self::getIdFromUuid($uuid, $model);
```
Comment (darrynten) [inline app/BaseModel.php]: You can remove the static and call `$this->getIdFromUuid` no need for it to be static

File: app/BaseModel.php [hunk 51]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    /**
+     * Get true ID from UUID for a given model
+     *
+     * @param $uuid
+     * @param $model
+     * @return mixed
+     */
+    private static function getIdFromUuid($uuid, $model)
```
Comment (darrynten) [inline app/BaseModel.php]: Per above, remove the static

File: app/Http/Controllers/UserController.php [hunk 52]
```diff
@@ -46,7 +58,8 @@ public function updateUserByUUID(Request $request, $userId)
         ]);
 
         if (Auth::user()->_id !== $userId) {
-            throw new \Exception('Illegal attempt to adjust another users details. The suspicious action has been logged.');
+            throw new \Exception('Illegal attempt to adjust another users details. '
+                . 'The suspicious action has been logged.');
```
Comment (darrynten) [inline app/Http/Controllers/UserController.php]: Concat goes at the end of the previous line

File: app/Http/Middleware/RolesMiddleware.php (hunk 30 above)
Comment (darrynten) [inline app/Http/Middleware/RolesMiddleware.php]: Unneeded

File: database/seeds/Roles.php [hunk 53]
```diff
@@ -0,0 +1,79 @@
...
+        ]);
+
+        Role::create([
+            'id' => 1,
+            '_id' => Uuid::generate(4),
+            'name' => 'system',
+            'is_active' => true,
+            'created_by' => 1,
+            'updated_by' => 1,
+        ]);
```
Comment (darrynten) [inline database/seeds/Roles.php:41]: Please reorder this and the one above so it goes id 1, 2, 3

File: database/seeds/Roles.php [hunk 54]
```diff
@@ -0,0 +1,79 @@
...
+
+        if (app()->environment('local')) {
+            DB::table('user_role')->delete();
+        }
+
+        UserRole::create([
```
Comment (darrynten) [inline database/seeds/Roles.php]: Docblock/comment for something this obscure, no idea what it's doing :D

**Review by darrynten — CHANGES_REQUESTED** [review]

File: README.md [hunk 55]
```diff
@@ -196,6 +196,31 @@ Make sure you set the appropriate variables in your .env
...
+- `GET` `/roles/`
+- `POST` `/roles/{name}`
+- `DELETE` `/roles/{roleId}`
+- `POST` `/roles/{roleId}/activate`
+- `POST` `/roles/{roleId}/deactivate`
+
+##Assigning roles to users
+- `GET` `/users/{id}/roles`
+- `POST` `/users/{id}/roles/assign/{role}`
+- `POST` `/users/{id}/roles/revoke/{role}`
```
Comment (darrynten) [inline README.md]: `roleId`?

Comment (fergusdixon) [reply README.md]: Oops

**Review by darrynten — APPROVED** [review]
Looking fantastic
